### PR TITLE
test: date tests added

### LINKED ISSUE #1549: Tests for Date conversion methods
We have had several issues regarding date formatting and timezone settings. This is a very crucial area for an event based application and thus must be ensured to be working correctly on all fronts. In order to do so, we must write unit tests so that no changes break the desired working of the application.

Tests should include:
- Correct conversion from ISO 8601 format to string representation
- Conformance to timezone setting

**Would you like to work on the issue?**

No. I am busy in preparations of my exams

**Target Milestone to solve the issue**

Till which [milestone](https://github.com/fossasia/open-event-android/milestones) will you be fixing the issue?

### PR DESCRIPTION
**Pull request by arp95**
Fixes #1549 

Changes:

Added tests for date testing.

Screenshots for the change: 
-

### CODE REVIEW

**Comment by arp95** [comment]
@iamareebjamal 
i still have to check the timezone stuff.but this is the way to go right?

**Comment by iamareebjamal** [comment]
Correct

**Comment by iamareebjamal** [comment]
@mahikaw More tests need to be added. This is not enough

**Comment by mahikaw** [comment]
@iamareebjamal yes I'm aware. I did it by mistake and couldn't find a way to undo the review. Please ignore.

**Comment by arp95** [comment]
yeah actually having my exams so not able to devote much time. working on tests part.

**Comment by arp95** [comment]
@iamareebjamal why do not merge tag? tests are not fine?

**Comment by iamareebjamal** [comment]
@arp95 For WIP PRs, do not merge tags tell maintainers to not try to merge the PR because it is work in progress. This saves time as they know they don't have to review it now

Someone may see check status and a correct test and merge the PR, so it's just a reminder, nothing bad for the author of PR. It doesn't reflect anything wrong with the PR

**Comment by arp95** [comment]
@iamareebjamal no i am not taking anything personal. i made changes and then the label was added after that.

**Comment by iamareebjamal** [comment]
Oh, I didn't know you had added more tests, let me take a look

Sorry

File: android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            readDate.setTimeZone(TimeZone.getTimeZone("GMT+04:01"));
+            Date gmt_date_3 = readDate.parse(gmt_time);
+            Assert.assertEquals("12:00:02", gmt_date_3);
+
+            //checking base case 00:00:00
+            readDate.setTimeZone(TimeZone.getTimeZone("GMT+16:01"));
+            Date gmt_Date_4 = readDate.parse(gmt_time);
+            Assert.assertEquals("00:00:02", gmt_Date_4);
+
+        }catch (ParseException e){
```
Comment (iamareebjamal) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: Add a space after curly brace
`} catch (`

Comment (arp95) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: sorry for delaying this for so long. having a look

Comment (arp95) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: @iamareebjamal i am confused over here. we use ISO8601Date.java class to test the correctness. but each function needs getTimeZone() which if i get it, i will have to use if-else statements to check the output, like if timezone is ist do this etc?

Comment (iamareebjamal) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: OK. So what you can do is increase a parameter in each method, taking in a timezone and create helper method for others that by default call getTimezone()

This generally happens while testing

Second option is using Mockito, which returns what you want when getTimezone() is called, but it will be a little difficult, so use first method

Comment (arp95) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: you mean adding more methods in ISO8601Date.java where i am passing timezone as a parameter?

Comment (iamareebjamal) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: Yes, and then use that general method in other methods as a helper one

I'm talking about function overloading using different parameter list

**Comment by arp95** [comment]
@iamareebjamal @Shailesh351 @mahikaw @shreyasnbhat please review.

File: android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import org.junit.Test;
+
+/**
+ * Created by arpitdec5 on 02-06-2017.
+ */
+
+public class DateTest {
+
+    @Test
+    public void checkDate() {
```
Comment (iamareebjamal) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: Ideally, there should be only one assert per test/function

Please break these into several tests and specify what it is expected to do
You can group similar assertions in one test

Look at this for reference - https://github.com/fossasia/open-event-orga-app/blob/master/app/src/test/java/org/fossasia/openevent/app/utils/UtilsTest.java

You can take a look at other test classes as well, there are plenty for reference

File: android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        //test of method getTimeZoneDateStringFromString()
+        ISO8601Date.setTimeZone("UTC");
+        ISO8601Date.setEventTimeZone("Asia/Singapore");
+        String dateString1 = "2017-06-02T07:59:10Z";
+        String dateString2 = "2017-06-02T20:00:10Z";
+        String dateString3 = "2017-06-03T00:00:10Z";
+
+        String predictedDateString1 = ISO8601Date.getTimeZoneDateStringFromString(dateString1);
+        String actualDateString1 = "Thu, 01 Jun 2017, 23:59, UTC";
+        Assert.assertEquals(predictedDateString1, actualDateString1);
```
Comment (iamareebjamal) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: The order should be this - `assert(actual, predicted)`

File: android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java (hunk 2 above)
Comment (iamareebjamal) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: This makes your tests run in parallel and hence, fast

**Comment by arp95** [comment]
@iamareebjamal i have to just break it into several tests right? the checking is fine or should i add more?
i think i covered most of the edge cases.

**Comment by iamareebjamal** [comment]
Looks good for now. I'll think of more use cases

File: android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        String day2 = "2017-06-02";
+        String predictedString;
+        String actualString;
+
+        predictedString = ISO8601Date.getTimeZoneDateStringFromStringForDayFragment(day1);
+        actualString = "1 Jun";
+        Assert.assertEquals(actualString, predictedString);
+
+        predictedString = ISO8601Date.getTimeZoneDateStringFromStringForDayFragment(day2);
+        actualString = "2 Jun";
```
Comment (iamareebjamal) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: This is actually expected and the above is actual

Please change others accordingly. Use `expected` instead of predicted to avoid confusion.

Also, make your test names show what they are expected to do. Like `shouldConvertLocalTimeZoneStringTo...`

In this way, you can break this tests into much smaller components. See the link I gave as a reference. A signle method can have multiple tests. It is easier to see what use case failed when debugging failed errors

**Comment by arp95** [comment]
@iamareebjamal is it fine now?

**Comment by iamareebjamal** [comment]
Looks good @arp95 but currently contains tests only for UTC and Singapore. Test for other combination of timezones can be added as well. The reason being that the application implementation may be hardcoded for UTC and singapore making the tests pass right now. You and I know that it isn't but in black box testing, we test without knowing about the implementation. 

These test conditions could be easily added : 
1. Same event timezone as local timezone -> Date should not change
2. Same event timezone and test for different local timezones
3. Same local timezone and changing event timezone

Use different methods to test each and also, no need to use many dates in the new tests now as the edge cases with many tests were already checked in the previous tests. We don't want overlapping tests

**Comment by arp95** [comment]
@iamareebjamal sorry for not seeing your post.
so what you mean is i can have more methods for testing having localtimezone as IST and event ones same and vice versa right? 
Also if i have to create the methods more how to name them? Is there a better way to name?

**Comment by arp95** [comment]
@iamareebjamal please check now.

File: android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+import org.junit.Test;
+
+/**
+ * Created by arpitdec5 on 02-06-2017.
+ */
+
+public class DateTest {
+
+    @Test
+    public void shouldConvertSingaporeDateStringToUTCDateString() {
```
Comment (iamareebjamal) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: Name format is correct but should not be particular like singapore or kolkata

It should be like shouldConvertLocalToInternational and then contain several tests on same string for different locales using same method.

For example, you are testing getTimeZoneDateStringFromString, so the test will only contain one date to be tested. And it will be tested on different timezones using the same method, so that if this particular test fails, we are sure that this particular feature of this method, namely changing the local date to international isn't working. All tests are good, just a little rearranging will do

One more thing. Keep the several date testing in one method for each function, like here you are using 4 strings, test them only once for a particular locale and this test will tell that the date conversion is correct. And when this test along with the above specified test (of timezone conversion) passes, we will know that both these functions of that method are working well

So, one containing 4 strings and testing on single locale - date conversion is correct
One containing single string and converting them to different locale or single string converting from different locale to local - timezone conversion is correct

This way each method will have 3 symmetrical tests which if failing will tell us that which component of the function is faulty

Comment (arp95) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: basically you are trying to say keep all the stuff in one method right?
one testing all 4 strings with specified timezone and eventTimeZone and the the single dateString testing with different timeZones?

Comment (iamareebjamal) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: No each method of Date class will have 3 tests, so 3 functions

Comment (iamareebjamal) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: 1 testing 4 dates for correct conversion, 2nd for local to international testing only one string but on 4 timezones, 3rd one international to local 1 string for 4 TZ

Comment (arp95) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: like for testing of getTimeZoneDateStringFromString(), i do following stuff:

1: use for date strings to check conversion from singapore time to utc
2: use two date strings to check conversion from singapore to amsterdam
3: use two date strings to check conversion from singapore to ist
4: use two date strings to check conversion from amsterdam to ist

you are saying to include all the above in one method right?

Comment (arp95) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: got it. means for each testing method 3 functions.

Comment (iamareebjamal) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: Yes, 3 tests methods for each date operation

Comment (iamareebjamal) [inline android/app/src/androidTest/java/org/fossasia/openevent/DateTest.java]: And keep the test names general

**Comment by arp95** [comment]
@iamareebjamal please check now.
